Ant script get_bundled.xml rewritten to Gradle

Crazyjavahacking opened the pull request:
Behavior changes
------------------
There are now 2 modes of TC artifact resolve:
1. **download TC artifacts from TC build ID (same behaviour as previously)**
    - ```
      ./gradlew clean getBundled
      Resolving TC build: Build{id=1545702, buildConfigurationId=Kotlin_1260_CompilerAllPlugins, buildNumber=1.2.60-release-76, status=SUCCESS, branch=BranchImpl(name=1.2.60, isDefault=true)}
      Downloading artifact: kotlin-plugin-1.2.60-release-IJ2017.3-1.zip
      ...
      ```
    - to override values from build script
      ```
      ./gradlew clean getBundled -PkotlinCompilerTcBuildId=1646860 -PideaVersion=162.1812.17 -PkotlinIdeaCompatibleVersionMinor=2017.3
      Resolving TC build: Build{id=1646860, buildConfigurationId=Kotlin_1270_CompilerAllPlugins, buildNumber=1.2.72-release-68, status=SUCCESS, branch=BranchImpl(name=1.2.70, isDefault=true)}
      Downloading artifact: kotlin-plugin-1.2.72-release-IJ2017.3-1.zip
      ...
      ```
2. **download TC artifacts from latest build of a specified version**
    - ```
      ./gradlew clean getBundled -PlastSuccessfulBuild
      Resolving TC build: Build{id=1617194, buildConfigurationId=Kotlin_1260_CompilerAllPlugins, buildNumber=1.2.61-release-102, status=SUCCESS, branch=BranchImpl(name=1.2.60, isDefault=true)}
      Downloading artifact: kotlin-plugin-1.2.61-release-IJ2017.3-3.zip
      ...
      ```
    - to override values from build script
      ```
      ./gradlew clean getBundled -PlastSuccessfulBuild -PkotlinCompilerVersion=1.3-M2 -PideaVersion=162.1812.17 -PkotlinIdeaCompatibleVersionMinor=2017.3
      Resolving TC build: Build{id=1590265, buildConfigurationId=Kotlin_13M2_CompilerAllPlugins, buildNumber=1.3-M2-release-219, status=SUCCESS, branch=BranchImpl(name=1.3-M2, isDefault=true)}
      Downloading artifact: kotlin-plugin-1.3-M2-release-IJ2017.3-1.zip
      ...
      ``` 

Resolving the artifacts from latest build
---------------------------------------
The resolve supports builds with 0 artifacts. In that case recursive search to the first build with all required artifacts will be performed.

Migration to Gradle
-------------------
- naming conventions & Gradle
    - property names do not use `a.b.c` naming convention (this would be interpreted as chained field access in Groovy's GString type) and is not used in Gradle in general
    - some of the properties were renamed
    - created proper Gradle task for each logical operation (downloading, extraction, repackaging)
    - added Gradle Wrapper 4.10.2
- general
    - relative paths replaced by absolute paths relative to Gradle project (so that changing working dir will not change behaviour)
    - Eclipse related Ant target was removed (as it was always skipped)
- downloading files
    - downloading test dependency (GSON) was replaced by downloading standard Maven artifact from Maven Cental
    - downloading of the files using HTTP (except GSON test dependency) is delegated to Ant's `get` task to retain backwards compatibility (HTTP moves, redirects, etc.)

Gradle task graph
------------------
![Imgur](https://i.imgur.com/zoRXVcQ.png)

Behavior comparison with Ant
------------------------------
- the Gradle execution creates the same filesystem changes as Ant execution with the following differences:
    1. generated `MANIFEST.MF` is slightly different (no metadata related to Ant: `Ant-Version` and `Created-By` keys) for the following files:
        - `file:/.../kotlin-bundled-compiler/lib/idea-formatter.jar!/META-INF/MANIFEST.MF`
        - `file:/.../kotlin-bundled-compiler/lib/opanapi-formatter.jar!/META-INF/MANIFEST.MF`
        - `file:/.../kotlin-bundled-compiler/lib/util-formatter.jar!/META-INF/MANIFEST.MF`

Crazyjavahacking committed:
Ant script get_bundled.xml rewritten to Gradle, added Gradle Wrapper 4.10.2, updated .gitignore

Kordyjan commented:
Migration to gradle is well done. Thank you.

However this still doesn't solve our main problem: build id of dependencies being hardcoded in our buildscript. 
It is ok for releases or nightly builds using released kotlin compiler versions. On the other hand, when we are doing integration with new, not yet released, version of compiler it is crucial for us to download artifacts from last successful build. I was thinking about some kind of command line parameter passed to buildscript. For example passing `-PkotlinBuild=Kotlin_130_CompilerAllPlugins` would cause overriding of hardcoded build id and downloading artifacts from last build from main branch of [1.3.0](https://teamcity.jetbrains.com/viewType.html?buildTypeId=Kotlin_130_CompilerAllPlugins&branch_Kotlin_130=%3Cdefault%3E&tab=buildTypeStatusDiv). It is not trivial because buildscript must be aware about suffixes like `-rc-151` or `-eap-43` in names of artifacts, but I believe it is doable using teamcity api.

Crazyjavahacking committed:
Complete rewrite of the TC artifacts handling:
1. artifacts now use Teamcity Rest Client library written in Kotlin
2. TC artifacts are now dynamically searched according to their file pattern

Crazyjavahacking committed:
Getting TC build by ID rewritten to already existing API TeamCityInstance.build(BuildId)

Crazyjavahacking committed:
Added integration tests to verify it is possible to download release, RC and EAP version of TC artifact when specified TC build ID

Crazyjavahacking committed:
Removed usages of kotlinCompilerTcBuildTypeId (which is now determined from kotlinCompilerVersion). Integration tests are now parameterized

Crazyjavahacking committed:
Added support for recursive build search when 0 artifacts are published for a TC build, added integration tests

Crazyjavahacking committed:
Integrated kotlin-rest-client with all required modifications (v. 1.5.0)

Crazyjavahacking committed:
Code refactored so that it can be reused in other projects

Crazyjavahacking committed:
Idea artifact functionality factored out to a separate HttpArtifactsResolver

Crazyjavahacking committed:
Integration tests for resolving TC build not containing any artifacts is now more strict

Crazyjavahacking commented:
This should contain the final changes with the following changes:
- added the dynamic resolve of artifacts
- integrated with TeamCity Rest client v 1.5.0 (which contains required changes)
- core factored to be reusable in other projects
- more strict integration tests
